smithy-model/src/main/java/software/amazon/smithy/model/knowledge/PropertyBindingIndex.java

```diff
@@ -0,0 +1,206 @@
+     * @return true if member shape maps to a property on the given resource
+     */
+    public boolean isMemberShapeProperty(ShapeId memberShapeId) {
+        return memberShapeToPropertyName.containsKey(memberShapeId);
+    }
+
+    /**
+     * Resolves and returns the output shape of an operation that contains the
+     * top-level resource bound properties. Handles adjustments made with
+     * @nestedProperties trait.
+     *
+     * @param operation operation to retrieve output properties shape for.
+     * @return the output shape of an operation that contains top-level resource
+     *  properties.
+     */
+    public Shape getOutputPropertiesShape(OperationShape operation) {
+        Model model = getModel();
+        return getPropertiesShape(operationIndex.getOutputMembers(operation).values(),
+                model.expectShape(operation.getOutputShape()));
+    }
+
+    /**
+     * Resolves and returns the input shape of an operation that contains the
+     * top-level resource bound properties. Handles adjustments made with
+     * @nestedProperties trait.
+     *
+     * @param operation operation to retrieve output properties shape for
+     * @return the input shape of an operation that contains top-level resource
+     *  properties.
+     */
+    public Shape getInputPropertiesShape(OperationShape operation) {
+        Model model = getModel();
+        return getPropertiesShape(operationIndex.getInputMembers(operation).values(),
+                model.expectShape(operation.getInputShape()));
+    }
+
+    /**
+     * Returns true if member is required to have an associated property mapping.
+     *
+     * @param memberShapeId the ShapeId of the member shape to check
+     * @return True if input/output member is required to have a property mapping.
+     */
+    public boolean doesMemberShapeRequireProperty(ShapeId memberShapeId) {
+        return !memberShapeDoesNotRequireProperty.getOrDefault(memberShapeId, false);
+    }
+
+    private Model getModel() {
+        return Objects.requireNonNull(model.get(), "The dereferenced WeakReference<Model> is null");
+    }
+
+    private Set<ShapeId> computeNotPropertyTraits() {
+        Model model = getModel();
+        //iteration populates non-traits into the set as well, but that doesn't have a negative
+        //impact on desired use.
+        return model.shapes()
+            .filter(shape -> shape.hasTrait(NotPropertyTrait.class))
+            .map(shape -> shape.toShapeId())
+            .collect(Collectors.toSet());
+    }
+
+    private Optional<String> getPropertyTraitName(MemberShape memberShape) {
+        return memberShape.getTrait(PropertyTrait.class).flatMap(PropertyTrait::getName);
+    }
+
+    private boolean doesNotRequireProperty(MemberShape memberShape) {
+        return memberShape.hasTrait(NotPropertyTrait.class)
+            || notPropertyMetaTraitSet.stream().filter(traitId -> memberShape.hasTrait(traitId))
+                .findAny().isPresent();
```

Comment:
`.filter(<predicate>).findAny().isPresent()` is replaceable with `.anyMatch(<predicate>)`